honeybadgermpc/progs/fixedpoint.py

```diff
@@ -0,0 +1,303 @@
+import logging
+import asyncio
+import time
+from honeybadgermpc.mpc import TaskProgramRunner
+from honeybadgermpc.elliptic_curve import Subgroup
+from honeybadgermpc.field import GF
+from honeybadgermpc.preprocessing import (
+    PreProcessedElements as FakePreProcessedElements,
+)
+from honeybadgermpc.progs.mixins.share_arithmetic import MixinConstants, BeaverMultiply
+
+
+config = {MixinConstants.MultiplyShare: BeaverMultiply()}
+
+
+# Fixed Point parameters
+
+F = 32  # The precision (binary bits)
+"""
+This implementation of the library is not completely hiding. This leaks information about the bits used in computation which is determinied by the security parameter Kappa.
+In particular, we leak O(1/(2^Kappa)) information theorotic bits per operation on a floating point secret.
+"""
+KAPPA = 32  # Statistical security parameter
+K = 64  # Total number of padding bits ()
+p = modulus = Subgroup.BLS12_381
+Field = GF(p)
+
+"""
+Library for fixed point operations:
+Throughout the code:
+`x` stands for private floating point number
+`k` is the total of number of bits we are representing
+
+"""
+
+
+# General (non MPC) fixed point functions
+""" Change a function to fixed point form. In general when we are dealing with fixed points, it
+is convert them back to fixed form.
+
+2.5 with f = 32 goes to 2.5* 2**32 which is an int.
+Note that this function always rounds down the error
+"""
+
+
+def to_fixed_point_repr(x, f=F):
+    return int(x * 2 ** f)
+
+
+"""
+Convert a number back from fixed point int form to python float.
+x: number to be converted.
+changes an int `x` to float value `x`/2**f deafult F=32
+"""
+
+
+def from_fixed_point_repr(x, k=K, f=F, signed=True):
+    x = x.value
+    if x >= 2 ** (k - 1) and signed:
+        x = -(p - x)
+
+    return float(x) / 2 ** f
+
+
+"""
+Convert x to a k-bit representation
+Least significant bit first
+"""
+
+
+def binary_repr(x, k):
+    def _binary_repr(v):
+        res = []
+        v = int(v)
+        for i in range(k):
+            res.append(v % 2)
+            v //= 2
+        return res
+
+    assert type(x) is int
+    return _binary_repr(x)
```

Comment:
Just to point out that an alternative way to get the binary repr of an integer can be done using the Python's builtin [`bin()`](https://docs.python.org/3/library/functions.html#bin) or [`format()`](https://docs.python.org/3/library/functions.html#format) such as in:

```python
def binary_repr(x, k):
    return [int(i) for i in f'{x:0{k}b}'[::-1]]
```

in a less compact form:


```python
def binary_repr(x, k):
    bin_str = f'{x:0{k}b}'
    reversed_bin_str = bin_str[::-1]
    return [int(i) for i in reversed_bin_str]
```

As for checking the type, the above will raise a `ValueError` if `x` is not an integer. If you prefer to catch the error and raise a different error:

```python
def binary_repr(x, k):
    try:
        bin_str = f'{x:0{k}b}'
    except ValueError:
        raise TypeError('x must be an integer')
     return [int(i) for i in bin_str[::-1]]
```

In general to check for the type of an argument `isinstance()` is preferred:

```python
assert isinstance(x, int)
```

See [PEP 8 Programming Recommendations](https://www.python.org/dev/peps/pep-0008/#programming-recommendations):

> Object type comparisons should always use [`isinstance()`](https://docs.python.org/3/library/functions.html#isinstance) instead of comparing types directly.
>
> Yes: if `isinstance(obj, int):`
> 
> No:  `if type(obj) is type(1):`